### PR TITLE
ShipmentScheduleEnqueuer - directly close the lock when done

### PR DESCRIPTION
**Pull request by metas-ts**
* ShipmentScheduleEnqueuer - don't wait for for the lock to be closed async later on
* If aquiring a lock fails, select and output the existing locks

### CODE REVIEW

File: backend/de.metas.async/src/main/java/de/metas/lock/api/impl/Lock.java [hunk 1]
```diff
@@ -85,18 +90,6 @@ public String toString()
 		return ObjectUtils.toString(this);
```
Comment (teosarca) [inline backend/de.metas.async/src/main/java/de/metas/lock/api/impl/Lock.java:90]: offtopic: consider getting rid of ObjectUtils.toString().
Replace it with `@ToString` and `@ToString.Exclude`.

File: backend/de.metas.async/src/main/java/de/metas/lock/api/impl/Lock.java [hunk 2]
```diff
@@ -49,9 +50,13 @@
 	private static final transient Logger logger = LogManager.getLogger(Lock.class);
 
 	/* package */CloseableReentrantLock mutex = new CloseableReentrantLock();
```
Comment (teosarca) [inline backend/de.metas.async/src/main/java/de/metas/lock/api/impl/Lock.java:52]: consider making it final.

File: backend/de.metas.async/src/main/java/de/metas/lock/exceptions/LockException.java [hunk 3]
```diff
@@ -86,7 +86,11 @@ protected final ITranslatableString buildMessage()
 		{
 			message.append("\n SQL Params: ").append(Arrays.toString(sqlParams));
 		}
-
+		if (existingLocks != null)
+		{
+			message.append("\n Existing Locks: ").append(existingLocks.toString());
```
Comment (teosarca) [inline backend/de.metas.async/src/main/java/de/metas/lock/exceptions/LockException.java:91]: `.toString()` call is not needed

File: backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java [hunk 4]
```diff
@@ -245,11 +263,92 @@ protected int performLockSQLInsert(final ILockCommand lockCommand, final List<Ob
...
+	private ImmutableList<ExistingLockInfo> retrieveExistingLocksForSelection(final int adTableId, @NonNull final PInstanceId pinstanceId)
+	{
+		return retrieveExistingLocksForWhereClause(
+				adTableId,
+				I_T_Lock.COLUMNNAME_Record_ID + " IN (select T_Selection_ID from T_Selection WHERE AD_PInstance_ID=" + pinstanceId.getRepoId() + ")",
+				null);
+	}
+
+	private ImmutableList<ExistingLockInfo> retrieveExistingLocksForWhereClause(
+			final int adTableId,
```
Comment (teosarca) [inline backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java:283]: use AdTableId

File: backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java [hunk 5]
```diff
@@ -245,11 +263,92 @@ protected int performLockSQLInsert(final ILockCommand lockCommand, final List<Ob
...
+						+ I_T_Lock.COLUMNNAME_IsAutoCleanup + ", "
+						+ I_T_Lock.COLUMNNAME_IsAllowMultipleOwners + ", "
+						+ " FROM " + I_T_Lock.Table_Name
+						+ " WHERE " + I_T_Lock.COLUMNNAME_AD_Table_ID + " = " + adTableId
+						+ " AND " + whereSql;
+		PreparedStatement pstmt = null;
+		ResultSet rs = null;
+		try
+		{
+			pstmt = DB.prepareStatement(sql, ITrx.TRXNAME_None);
```
Comment (teosarca) [inline backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java:301]: use org.compiere.util.DB#retrieveRows(java.lang.CharSequence, java.util.List<java.lang.Object>, org.compiere.util.DB.ResultSetRowLoader<T>)

File: backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java [hunk 6]
```diff
@@ -245,11 +263,92 @@ protected int performLockSQLInsert(final ILockCommand lockCommand, final List<Ob
...
+		{
+			throw AdempiereException.wrapIfNeeded(e);
+		}
+		finally
+		{
+			DB.close(rs, pstmt);
+		}
+	}
+
+	private ExistingLockInfo extractExistingLockInfo(@NonNull final ResultSet rs) throws SQLException
```
Comment (teosarca) [inline backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java:323]: static

File: backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java [hunk 7]
```diff
@@ -245,11 +263,92 @@ protected int performLockSQLInsert(final ILockCommand lockCommand, final List<Ob
...
+		return ExistingLockInfo
+				.builder()
+				.lockId(rs.getInt(I_T_Lock.COLUMNNAME_T_Lock_ID))
+				.ownerName(I_T_Lock.COLUMNNAME_Owner)
+				.lockedRecord(TableRecordReference.of(
+						rs.getInt(I_T_Lock.COLUMNNAME_AD_Table_ID),
+						rs.getInt(I_T_Lock.COLUMNNAME_Record_ID)
+				))
+				.allowMultipleOwners(rs.getBoolean(I_T_Lock.COLUMNNAME_IsAllowMultipleOwners))
+				.autoCleanup(rs.getBoolean(I_T_Lock.COLUMNNAME_IsAutoCleanup))
```
Comment (teosarca) [inline backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java:334]: i think u want to use StringUtils.toBoolean(rs.getString(...))

**Review by teosarca — CHANGES_REQUESTED** [review]
see my comments

File: backend/de.metas.async/src/main/java/de/metas/lock/api/impl/Lock.java (hunk 1 above)
Comment (metas-ts) [reply backend/de.metas.async/src/main/java/de/metas/lock/api/impl/Lock.java:90]: done (note: didn't find anything that i wanted to exclude)

File: backend/de.metas.async/src/main/java/de/metas/lock/api/impl/Lock.java (hunk 2 above)
Comment (metas-ts) [reply backend/de.metas.async/src/main/java/de/metas/lock/api/impl/Lock.java:52]: done

File: backend/de.metas.async/src/main/java/de/metas/lock/exceptions/LockException.java (hunk 3 above)
Comment (metas-ts) [reply backend/de.metas.async/src/main/java/de/metas/lock/exceptions/LockException.java:91]: yes it is :-P

File: backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java (hunk 4 above)
Comment (metas-ts) [reply backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java:283]: done

File: backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java (hunk 5 above)
Comment (metas-ts) [reply backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java:301]: aweome, thx!

File: backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java (hunk 6 above)
Comment (metas-ts) [reply backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java:323]: done

File: backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java (hunk 7 above)
Comment (metas-ts) [reply backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java:334]: you read my mind :blush:

File: backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java [hunk 8]
```diff
@@ -85,7 +88,10 @@ private static final String toSqlParam(final Object param, final List<Object> sq
 		}
 	}
 
-	private final void appendTableRecordWhereClause(final TableRecordReference record, final StringBuilder sql, final List<Object> sqlParams)
+	private void appendTableRecordWhereClause(
+			@NonNull final TableRecordReference record,
+			@NonNull final StringBuilder sql,
+			@NonNull final List<Object> sqlParams)
 	{
 		final int adTableId = record.getAD_Table_ID();
```
[the remark this answers is not in the document]
Comment (metas-ts) [reply backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java:96]: done

File: backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java (hunk 5 above)
Comment (teosarca) [reply backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java:301]: @metas-ts  can u pls also fix this, just saw it 🤦‍♂️ 
image

File: backend/de.metas.async/src/main/java/de/metas/lock/api/impl/AbstractLockDatabase.java [hunk 9]
```diff
@@ -55,23 +56,22 @@
 	protected final transient Logger logger = LogManager.getLogger(getClass());
 
 	/** Asserts given lock owner is a valid owner to be used on for Locks */
-	protected static final void assertValidLockOwner(final LockOwner lockOwner)
+	protected static void assertValidLockOwner(@NonNull final LockOwner lockOwner)
 	{
 		// NOTE: LockOwner.ANY is not tolerated because that's a filter criteria and not a LockOwner that we could use for assigning
 
-		Check.assumeNotNull(lockOwner, "Lock owner shall not be null");
 		Check.assumeNotNull(lockOwner.isRealOwnerOrNoOwner(), "Lock owner shall be real owner or no owner but it was {}", lockOwner);
```
Comment (teosarca) [inline backend/de.metas.async/src/main/java/de/metas/lock/api/impl/AbstractLockDatabase.java:63]: i think here we need Check.assume

File: backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java (hunk 5 above)
Comment (metas-ts) [reply backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java]: done

File: backend/de.metas.async/src/main/java/de/metas/lock/api/impl/AbstractLockDatabase.java (hunk 9 above)
Comment (metas-ts) [reply backend/de.metas.async/src/main/java/de/metas/lock/api/impl/AbstractLockDatabase.java]: done

File: backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java [hunk 10]
```diff
@@ -245,15 +267,79 @@ protected int performLockSQLInsert(final ILockCommand lockCommand, final List<Ob
...
+		final String sql =
+				"SELECT " + I_T_Lock.COLUMNNAME_AD_Table_ID + ", "
+						+ I_T_Lock.COLUMNNAME_T_Lock_ID + ", "
+						+ I_T_Lock.COLUMNNAME_Record_ID + ", "
+						+ I_T_Lock.COLUMNNAME_Owner + ", "
+						+ I_T_Lock.COLUMNNAME_IsAutoCleanup + ", "
+						+ I_T_Lock.COLUMNNAME_IsAllowMultipleOwners
+						+ " FROM " + I_T_Lock.Table_Name
+						+ " WHERE " + I_T_Lock.COLUMNNAME_AD_Table_ID + " = " + adTableId.getRepoId()
+						+ " AND " + whereSql;
```
Comment (metas-ts) [inline backend/de.metas.async/src/main/java/de/metas/lock/spi/impl/SqlLockDatabase.java:304]:
```suggestion
						+ " AND ( " + whereSql + " )";
```

**Review by metas-ts — COMMENTED** [review]
lgtm

**Pull request merged**